scipy/sparse/linalg/eigen/arpack/arpack.py

```diff
@@ -1231,11 +1231,29 @@ def eigs(A, k=6, M=None, sigma=None, which='LM', v0=None,
             import warnings
             warnings.warn('M does not have the same type precision as A. '
                           'This may adversely affect ARPACK convergence')
+
     n = A.shape[0]
 
-    if k <= 0 or k >= n:
-        raise ValueError("k=%d must be between 1 and ndim(A)-1=%d"
-                         % (k, n - 1))
+    if k <= 0:
+        raise ValueError("k=%d must be greater than 0." % k)
+
+    if k >= n - 1:
+        if issparse(A):
+            raise TypeError("Cannot use scipy.linalg.eig for sparse A with "
+                            "k >= N - 1. Use scipy.linalg.eig(A.toarray()) or"
+                            " reduce k.")
+        if isinstance(A, LinearOperator):
+            raise TypeError("Cannot use scipy.linalg.eig for LinearOperator "
+                            "A with k >= N - 1.")
+        if isinstance(M, LinearOperator):
+            raise TypeError("Cannot use scipy.linalg.eig for LinearOperator "
+                            "M with k >= N - 1.")
+
+        import warnings
```

Comment:
You just need to import `warnings` module at the top of the file and you can use it in any function. Also with 

    from .misc import LinAlgWarning

you can raise a `LinAlgWarning` which I guess is more relevant instead of a `UserWarning`. Also you already know the values of k and N so the message can use those values

        warnings.warn("{0} >= {1} for {1}x{1} square matrix. "
                      "Using scipy.linalg.eigh instead.".format(k, N),
                      LinAlgWarning, stacklevel=3)

Reply:
I cant find `LinAlgWarning` because this branch has not been updated to v1.1. Should I merge this branch with the latest master branch to get `LinAlgWarning`?

Reply:
Then I guess the regular RuntimeWarning would suffice. At some point we will sweep all the warnings anyhow.